R/gh_token.R

```diff
@@ -46,8 +46,17 @@
 gh_token <- function(api_url = NULL) {
   api_url <- api_url %||% default_api_url()
   stopifnot(is.character(api_url), length(api_url) == 1)
+  host_url <- get_hosturl(api_url)
+  # Session credentials take precedence over static credentials.
+  session <- get_connect_session()
+  if (!is.null(session)) {
```

Comment:
It took me a minute to realise that we don't need an explicit check for connectcreds here because that occurs in `get_connect_session()`. But I suspect that suggests these functions need to be factored slightly differently?

Maybe more like:

```R
if (has_connectcreds() && in_shinyapp()) { 
  session <- get_connect_session()
  token <- connectcreds::connect_viewer_token(session, host_url)
  if (!is.null(token)) {
    return(gh_pat(token$access_token))
  }
}
```

?

Reply:
The current form is largely because I'm struggling with how error propagation should happen here.

Right now `connect_viewer_token()` throws errors while `has_viewer_token()` does not.

The trouble is that we can't assume that just because a Shiny app is deployed to Connect the publisher wants to use viewer-based credentials. So clearly it can't be fatal if there are no credentials available, and that's why I'm using `has_viewer_token()`. But at the same time, it feels bad to swallow errors like "no OAuth integration is assigned to this app" that would probably help authors debug why their app isn't working.

Currently I have a `connectcreds_debug` option that works like `gargle`'s "debug" mechanism, emitting the errors as messages instead. But it's not on by default, so folks would probably only learn about it through the documentation... limiting its usefulness. Maybe `connectcreds` should log debug messages by default if it detects we're on Connect?

Do you have any thoughts on error reporting?

Reply:
It feels like we need some explicit signal from the user that they want to use viewer-based auth when deployed?

Reply:
We had that when we had an explicit `session` parameter :disappointed:.

Maybe some of this is just because `gh_token()` doesn't issue an error when there are no credentials. In `odbc` and `elmer` I notice that we can say things like "No credentials available ... by the way, you're running in Connect? Try turning on viewer-based credentials, see _link_".

Reply:
So maybe we're better off switching back to that UI? 

Maybe `gh_token()` should error? But even if that is the correct behaviour there likely to be a bunch of downstream changes because of that.